core/src/main/java/org/elasticsearch/common/text/UTF8SortedAsUnicodeComparator.java

```diff
@@ -18,60 +18,76 @@
-            int bUpto = b.arrayOffset();
-
-            final int aStop = aUpto + Math.min(a.length(), b.length());
-            while (aUpto < aStop) {
-                int aByte = aBytes[aUpto++] & 0xff;
-                int bByte = bBytes[bUpto++] & 0xff;
-
-                int diff = aByte - bByte;
-                if (diff != 0) {
-                    return diff;
-                }
-            }
+        try {
+            // we use the iterators since it's a 0-copy comparison where possible!
+            final BytesRefIterator aIter = a.iterator();
+            final BytesRefIterator bIter = b.iterator();
+            final long lengthToCompare = Math.min(a.length(), b.length());
+            BytesRef aRef = aIter.next();
+            BytesRef bRef = bIter.next();
+            if (aRef != null && bRef != null) { // do we have any data?
+                if (aRef.length == a.length() && bRef.length == b.length()) { // is it only one array slice we are comparing?
+                    return aRef.compareTo(bRef);
+                } else {
+                    for (int i = 0; i < lengthToCompare;) {
+                        if (aRef.length == 0) {
+                            aRef = aIter.next();
+                            if (aRef == null) {
+                                break;
+                            }
+                        }
+                        if (bRef.length == 0) {
+                            bRef = bIter.next();
+                            if (bRef == null) {
+                                break;
+                            }
+                        }
+                        final int aLength = aRef.length;
+                        final int bLength = bRef.length;
+                        final int length = Math.min(aLength, bLength); // shrink to the same length and use the fast compare in lucene
+                        aRef.length = bRef.length = length;
+                        // now we move to the fast comparison - this is the hot part of the loop
+                        int diff = aRef.compareTo(bRef);
+                        aRef.length = aLength;
+                        bRef.length = bLength;
 
-            // One is a prefix of the other, or, they are equal:
-            return a.length() - b.length();
-        } else {
-            final byte[] aBytes = a.toBytes();
-            int aUpto = 0;
-            final byte[] bBytes = b.toBytes();
-            int bUpto = 0;
-
-            final int aStop = aUpto + Math.min(a.length(), b.length());
-            while (aUpto < aStop) {
-                int aByte = aBytes[aUpto++] & 0xff;
-                int bByte = bBytes[bUpto++] & 0xff;
-
-                int diff = aByte - bByte;
-                if (diff != 0) {
-                    return diff;
+                        if (diff != 0) {
+                            return diff;
+                        }
+                        advance(aRef, length);
+                        advance(bRef, length);
```

Comment:
I think it can be dangerous that we modify the BytesRef returned by the iterator as it requires the iterator to set offset and length at every iteration, which sose impls might not have to.